ttyplot.c

```diff
@@ -289,9 +290,99 @@ void redraw_screen(const char * errstr) {
     refresh();
 }
 
+// Handle a single value from the input stream.
+// Return whether we got a full data record.
+int handle_value(double value)
+{
+    static double saved_value;
+    static int saved_value_valid = 0;
+
+    // First value of a 2-value record: save it for later.
+    if (two && !saved_value_valid) {
+        saved_value = value;
+        saved_value_valid = 1;
+        return 0;
+    }
+
+    // Otherwise we have a full record.
+    n = (n+1) % plotwidth;
+    if (two) {
+        values1[n] = saved_value;
+        values2[n] = value;
+        saved_value_valid = 0;
+    } else {
+        values1[n] = value;
+    }
+    if (rate)
+        td = derivative(&values1[n], two ? &values2[n] : NULL, &now);
+    return 1;
+}
+
+// Parse the input data. Return the number of bytes consumed.
+size_t parse_input(char *buffer, size_t length)
+{
+    static const char delimiters[] = " \t\r\n";  // white space
+
+    // Find the last delimiter.
+    int end;
+    for (end = length - 1; end >= 0; end--) {
+        if (strchr(delimiters, buffer[end]))  // found
+            break;
+    }
+    if (end < 0)  // no delimiter
+        return 0;
+    buffer[end] = '\0';
+
+    // Tokenize and parse.
+    int records = 0;  // number or records found
+    char *str = buffer;
+    char *token;
+    while ((token = strtok(str, delimiters)) != NULL) {
+        records += handle_value(strtod(token, NULL));
+        str = NULL;  // tell strtok() to stay on the same string
+    }
+    v += records;
+    if (records > 0)
+        redraw_screen(errstr);
+    return end + 1;
+}
+
+// Handle an "input ready" event, where only a single read() is guaranteed to not block.
+// Return whether the input stream got closed.
+int handle_input(void)
+{
+    static char buffer[4096];
+    static size_t buffer_pos = 0;
+
+    // Buffer incoming data.
+    int bytes_read = read(STDIN_FILENO, buffer + buffer_pos, sizeof buffer - 1 - buffer_pos);
+    if (bytes_read < 0) {  // read error
+        if (errno == EINTR || errno == EAGAIN)  // we should try again later
+            return 0;
+        errstr = strerror(errno);  // other errors are considered fatal
+        return 1;
+    }
+    if (bytes_read == 0) {
+        errstr = "input stream closed";
+        buffer[buffer_pos++] = '\n';  // attempt to extract one last value
+        parse_input(buffer, buffer_pos);
+        redraw_screen(errstr);  // redraw to display the error message
```

Comment:
(same as above)
If `handle_input` is about handling input, I would like to argue that it calling `redraw_screen` is out of place here because it violates the contract of what is the job of function `handle_input`. Can we pull that call out some way? I believe all calls to `redraw_screen` would ideally be from the `main` function.

Reply:
Same as above.